### ISSUE TITLE
Support mutable layers

### ISSUE THREAD

**Issue by ematejska** [tensorflow/swift-apis#893]
Imported from JIRA, filed by Chris Lattner here: https://bugs.swift.org/browse/TF-481

BatchNorm implementations are pretty suboptimal right now, we need to figure out a path to make it so they can be implemented in a straight-forward way.

I'm not sure the best path on this, but it could be something like:

1) by-ref parameters
2) by-ref layers
3) mutating func call()
4) standardizing a ref wrapper

Relatd to: SR-12148 Make differentiation work with class types

**Comment by dan-zheng** [comment]
I think we can close this issue.

Differentiation has support for class types now (SR-12148), with some missing functionality and known compiler crashers.

It is possible to declare a `Layer`-conforming class type, and we _could_ implement `BatchNorm` that way instead of using `Parameter<Tensor<Scalar>>` class-typed stored properties.
